server/src/main/java/org/elasticsearch/action/search/CCSSingleCoordinatorSearchProgressListener.java

```diff
@@ -0,0 +1,315 @@
+
+        for (Map.Entry<String, Integer> entry : totalByClusterAlias.entrySet()) {
+            String clusterAlias = entry.getKey();
+            int successfulCount = entry.getValue().intValue();
+
+            AtomicReference<SearchResponse.Cluster> clusterRef = clusters.getCluster(clusterAlias);
+            boolean swapped;
+            do {
+                SearchResponse.Cluster curr = clusterRef.get();
+                SearchResponse.Cluster.Status status = curr.getStatus();
+                if (status != SearchResponse.Cluster.Status.RUNNING) {
+                    // don't swap in a new Cluster if the final state has already been set
+                    break;
+                }
+                TimeValue took = new TimeValue(timeProvider.buildTookInMillis());
+                int successfulShards = successfulCount + curr.getSkippedShards();
+                assert successfulShards + curr.getFailedShards() == curr.getTotalShards()
+                    : "successfulShards("
+                        + successfulShards
+                        + ") + failedShards("
+                        + curr.getFailedShards()
+                        + ") != totalShards ("
+                        + curr.getTotalShards()
+                        + ')';
+                if (curr.isTimedOut() || successfulShards < curr.getTotalShards()) {
+                    status = SearchResponse.Cluster.Status.PARTIAL;
+                } else {
+                    assert successfulShards == curr.getTotalShards()
+                        : "successful (" + successfulShards + ") should equal total(" + curr.getTotalShards() + ") if get here";
+                    status = SearchResponse.Cluster.Status.SUCCESSFUL;
+                }
+                SearchResponse.Cluster updated = new SearchResponse.Cluster.Builder(curr).setStatus(status)
+                    .setSuccessfulShards(successfulShards)
+                    .setTook(took)
+                    .build();
+                swapped = clusterRef.compareAndSet(curr, updated);
+            } while (swapped == false);
+        }
+    }
+
+    /**
+     * Executed when a shard returns a fetch result.
+     *
+     * @param shardIndex The index of the shard in the list provided by {@link SearchProgressListener#onListShards})}.
+     */
+    @Override
+    public void onFetchResult(int shardIndex) {}
+
+    /**
+     * Executed when a shard reports a fetch failure.
+     *
+     * @param shardIndex The index of the shard in the list provided by {@link SearchProgressListener#onListShards})}.
+     * @param shardTarget The last shard target that thrown an exception.
+     * @param exc The cause of the failure.
+     */
+    @Override
+    public void onFetchFailure(int shardIndex, SearchShardTarget shardTarget, Exception exc) {}
+
+    private Map<String, Integer> partitionCountsByClusterAlias(Stream<SearchShard> shards) {
```

Comment:
I wonder if we shouldn't be a little more careful here with performance, the shards lists can become quite long?
Why not do something simple with predictable performance, like:

```diff
diff --git a/server/src/main/java/org/elasticsearch/action/search/CCSSingleCoordinatorSearchProgressListener.java b/server/src/main/java/org/elasticsearch/action/search/CCSSingleCoordinatorSearchProgressListener.java
index b13abcb8ce4..00567bcd712 100644
--- a/server/src/main/java/org/elasticsearch/action/search/CCSSingleCoordinatorSearchProgressListener.java
+++ b/server/src/main/java/org/elasticsearch/action/search/CCSSingleCoordinatorSearchProgressListener.java
@@ -16,11 +16,11 @@ import org.elasticsearch.search.aggregations.InternalAggregations;
 import org.elasticsearch.search.query.QuerySearchResult;
 import org.elasticsearch.transport.RemoteClusterAware;
 
+import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
+import java.util.Objects;
 import java.util.concurrent.atomic.AtomicReference;
-import java.util.stream.Collectors;
-import java.util.stream.Stream;
 
 /**
  * Use this progress listener for cross-cluster searches where a single
@@ -56,9 +56,10 @@ public class CCSSingleCoordinatorSearchProgressListener extends SearchProgressLi
         this.timeProvider = timeProvider;
 
         // Partition by clusterAlias and get counts
-        Map<String, Integer> skippedByClusterAlias = partitionCountsByClusterAlias(skipped.stream());
+        Map<String, Integer> skippedByClusterAlias = partitionCountsByClusterAlias(skipped);
         // the 'shards' list does not include the shards in the 'skipped' list, so combine counts from both to get total
-        Map<String, Integer> totalByClusterAlias = partitionCountsByClusterAlias(Stream.concat(shards.stream(), skipped.stream()));
+        Map<String, Integer> totalByClusterAlias = partitionCountsByClusterAlias(shards);
+        skippedByClusterAlias.forEach((cluster, count) -> totalByClusterAlias.merge(cluster, count, Integer::sum));
 
         for (Map.Entry<String, Integer> entry : totalByClusterAlias.entrySet()) {
             String clusterAlias = entry.getKey();
@@ -195,7 +196,7 @@ public class CCSSingleCoordinatorSearchProgressListener extends SearchProgressLi
      */
     @Override
     public void onPartialReduce(List<SearchShard> shards, TotalHits totalHits, InternalAggregations aggs, int reducePhase) {
-        Map<String, Integer> totalByClusterAlias = partitionCountsByClusterAlias(shards.stream());
+        Map<String, Integer> totalByClusterAlias = partitionCountsByClusterAlias(shards);
 
         for (Map.Entry<String, Integer> entry : totalByClusterAlias.entrySet()) {
             String clusterAlias = entry.getKey();
@@ -247,7 +248,7 @@ public class CCSSingleCoordinatorSearchProgressListener extends SearchProgressLi
             return;
         }
 
-        Map<String, Integer> totalByClusterAlias = partitionCountsByClusterAlias(shards.stream());
+        Map<String, Integer> totalByClusterAlias = partitionCountsByClusterAlias(shards);
 
         for (Map.Entry<String, Integer> entry : totalByClusterAlias.entrySet()) {
             String clusterAlias = entry.getKey();
@@ -306,10 +307,11 @@ public class CCSSingleCoordinatorSearchProgressListener extends SearchProgressLi
     @Override
     public void onFetchFailure(int shardIndex, SearchShardTarget shardTarget, Exception exc) {}
 
-    private Map<String, Integer> partitionCountsByClusterAlias(Stream<SearchShard> shards) {
-        return shards.collect(Collectors.groupingBy(shard -> {
-            String clusterAlias = shard.clusterAlias();
-            return clusterAlias != null ? clusterAlias : RemoteClusterAware.LOCAL_CLUSTER_GROUP_KEY;
-        }, Collectors.reducing(0, e -> 1, Integer::sum)));
+    private Map<String, Integer> partitionCountsByClusterAlias(List<SearchShard> shards) {
+        final Map<String, Integer> res = new HashMap<>();
+        for (SearchShard shard : shards) {
+            res.merge(Objects.requireNonNullElse(shard.clusterAlias(), RemoteClusterAware.LOCAL_CLUSTER_GROUP_KEY), 1, Integer::sum);
+        }
+        return res;
     }
 }
```

Then we don't have to iterate the skipped shards twice?

As a side note, we could (and maybe should for CCS in particular?) just pass iterators (or just the group shards iterators) into this method instead of collecting lists upstream in a follow up I think :)

Reply:
Thanks. I see what you meant earlier now. I've updated to your recommendation (next push).

Reply:
> As a side note, we could (and maybe should for CCS in particular?) just pass iterators (or just the group shards iterators) into this method instead of collecting lists upstream in a follow up I think :)

I might be missing what you mean, but I don't think I that fits here. If you mean pass GroupShardsIterator into the SearchProgressListener method - that interface was already defined before this recent CCS work. So do you mean to refactor that and its various implementations?

One potential problem with that is that the GroupShardsIterator doesn't guarantee that it knows the clusterAlias the shard belongs to, while the `SearchShard` object does. So that would require some additional refactoring or restructuring.

Or do you mean in the Cluster objects? But the Cluster objects just wants counts (and failure messages).

Reply:
I looked into the possibility of changing the `SearchProgressListener` to take GroupShardsIterator. Changing the `notifyListShards/onListShards` methods would be straightforward and be more efficient (since `AbstractSearchAsyncAction` wouldn't have to convert from `GroupShardsIterator` to `List<SearchShard>`).

But several other `SearchProgressListener` methods also take `List<SearchShard>` such as `notifyPartialReduce` and `notifyFinalReduce` which are called from the `QueryPhaseResultConsumer`, which does not operate on `GroupShardsIterator` / `SearchShardIterator`. 

Converting to SearchShardIterator could be more difficult because it requires a lot more information that what is available in later phases of search (?)

So it looks to me like `List<SearchShard>` was chosen as the simplest common interface across all the forms of "shard search actions" across the search phases.

Reply:
Yea that makes sense, oh well we can look into optimising this some other time :) It's not super important in the grand scheme of things today, but should we ever get more memory efficient overall this one would be a non-trivial contributor for larger searches potentially :)